### ISSUE TITLE
F# EF and type providers for SQL

### ISSUE THREAD

**Issue by jkone27** [dotnet/efcore#27307]
Using[ F# ](https://docs.microsoft.com/en-us/dotnet/fsharp/what-is-fsharp)and EF core together :) :) :) :) 

C# is amazing, and F# and C# work amazingly together!

Extra usage also: F# (other than compiled like C#) can be used in scripting/REPL mode (a bit like python), so db access would also made simple with .fsx scripts

make this project part of EF core ?!?! or contribute to it or have a similar F# ef-compatible alternative
https://fsprojects.github.io/SQLProvider/ 

i know i can reference a C# assembly, but F# provided types would give extra features to dotnet developers!
and the learning curve for only using SQL type providers is really not that steep at all!

I am just an F# enthusiast and F# works perfectly fine together with C#, just type providers is a unique F# feature from the compiler which[ also won an ACM (touring award committee) mention...](https://fsprojects.github.io/FSharp.Data/)

you can create an fsx file and run it from within VS2022 or vscode without creating any solutions at all!

C# developers culd just have a Repository as F# project, and invoke it from C# code whenever needed!

```fsharp
#r "nuget: SQLProvider"

let [<Literal>] dbVendor = 
   Common.DatabaseProviderTypes.MSSQLSERVER
let [<Literal>] connString = 
   "Server=localhost;Database=test;User Id=test;Password=test"

type SqlProvidedType =
    SqlDataProvider<
        dbVendor,
        connString>
        
  let dev_ctx () = 
      SqlProvidedType.GetDataContext()
      
   let prod_ctx prodConnectionString = 
      SqlProvidedType.GetDataContext prodConnectionString

//have fun: https://fsprojects.github.io/SQLProvider/core/general.html
let myDevCtx = dev_ctx()

// do your queries or commands here...

```

what about extending or integrating, working together with https://fsprojects.github.io/SQLProvider/ 
having an EF compatible provided type would be amazing :)

**Comment by roji** [comment]
@jkone27 I'm no expert, but AFAIK F# type providers shine where the type information is external to the application. For example, if you have a database, a type provider can discover it's schema and expose types based on the tables it finds; this is what SQLProvider does to my understanding.

EF Core, on the other hand relies on having a code model (as existnig .NET types) which represent database tables, and which are generally maintained and customized by users. So I'm not sure what role an F# type provider would have here, given that everything is already modeled in code inside the application. In addition, EF Core and SQLProvider both have their own LINQ providers, with each being a specific implementation; I'm not sure how they would mix.

To summarize, EF Core and SQLProvider are indeed similar in that they provide database access, but the approaches are quite different. I'm not sure exactly what kind of integration you'd like to see here - more details may help clarify that.

**Comment by jkone27** [comment]
where EF-core needs compile time generation to build types from db and vice-versa, but keeps them as separate versioned classes/types, 
an F# type provider version of EF-core could give users instant feedback (at edit time) and no extra code to handle or version in the repo for the data layer.

A classic "utility" is migrations, when applying a migration script at db level, even locally, the provided type already will have the new types, so code have to be adjusted to compile, making also some of the app integration tests nor required/necessary in this way

I was just picturing that since F# has kind of a "not awesome experience with EF" as programmability, that maybe a similar Provided EF compatible DbContext in F# (not using erased types but full types) would be much nicer and adopted.

 where also all unit tests, perf and integration tests are run for the usual EF-core library, so would guaranty a high level of quality and always "stay up to date probably" with the rest of the framework...
 
Also this would make EF-core accessible in dotnet scripts/scripting where F# is gaining usability and is very nice and immediate to use , just a thought!
 
and being able to use that from application code in C#, just invoking the F# assembly, would make it appetible also in mixed projects. just an idea ofc, thank you so much for your time for replying 👍 !!!


Cheers

**Comment by roji** [comment]
> where EF-core needs compile time generation to build types from db and vice-versa, but keeps them as separate versioned classes/types, an F# type provider version of EF-core could give users instant feedback (at edit time) and no extra code to handle or version in the repo for the data layer.

EF Core does not do compile-time generation of types from the database, and there are no plans for making it do so. EF Core does allow [scaffolding](https://docs.microsoft.com/en-us/ef/core/managing-schemas/scaffolding?tabs=dotnet-core-cli) an existing database into a .NET code model, but this is an explicit design-time gesture, and not a compile-time thing. Note that EF Core creates a C# (or F#) code model of your tables, which can later be customized by the user; this is quite different from creating .NET types (which aren't source code).

> I was just picturing that since F# has kind of a "not awesome experience with EF" as programmability, that maybe a similar Provided EF compatible DbContext in F# (not using erased types but full types) would be much nicer and adopted.

I definitely agree that the F# EF Core experience could be improved (e.g. immutable type support, F# records...), but I don't really think that type providers have anything to do with it.

> Also this would make EF-core accessible in dotnet scripts/scripting where F# is gaining usability and is very nice and immediate to use , just a thought!

Note once an EF model has been written (or scaffolded) in source code - F# or C# - EF Core can easily be used in F# REPL (or, for that matter in C# REPL - [see this blog post](https://devblogs.microsoft.com/dotnet/net-core-with-juypter-notebooks-is-here-preview-1/) for EF Core being used with Jupyter Notebooks, for example).
